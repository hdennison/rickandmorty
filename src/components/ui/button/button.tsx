import * as React from "react";
import { Slot } from "@radix-ui/react-slot";
import { cva, type VariantProps } from "class-variance-authority";

import { cn } from "../utils";
import { SearchIcon } from "../icon/icon";

const buttonVariants = cva(
  "flex items-center gap-2 cursor-pointer px-4 py-2 rounded-md justify-center",
  {
    variants: {
      variant: {
        primary: "bg-[#ddd] border-[#333] hover:bg-[#ccc]",
        secondary: "border-1 border-[gray] hover:bg-[#f0f0f0] text-[gray]",
      },
    },
    defaultVariants: {
      variant: "primary",
    },
  },
);

interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  asChild?: boolean;
}

interface ButtonComponent
  extends React.ForwardRefExoticComponent<
    ButtonProps & React.RefAttributes<HTMLButtonElement>
  > {
  Icon: typeof ButtonIcon;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : "button";
    return (
      <Comp
        className={cn(buttonVariants({ variant, className }))}
        ref={ref}
        {...props}
      />
    );
  },
) as ButtonComponent;

Button.displayName = "Button";

const icons = {
  search: SearchIcon,
};

const ButtonIcon = ({ icon }: { icon: keyof typeof icons }) => {
  const Comp = icons[icon];

  return <Comp size={16} />;
};

Button.Icon = ButtonIcon;

export { Button, buttonVariants };
