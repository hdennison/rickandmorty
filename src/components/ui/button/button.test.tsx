import { Button } from "@/components/ui/button/button";
import { describe, expect, test } from "vitest";
import { render } from "vitest-browser-react";

const variants = ["default", "primary", "secondary"] as const;

describe("variants", () => {
  variants.forEach((variant) => {
    test(`${variant} variant`, () => {
      const props = variant === "default" ? {} : { variant };

      const { getByText } = render(<Button {...props}>Click me!</Button>);

      const button = getByText("Click me!").element();
      // Snapshots are not the best solution, it woould be much better to set up Storybook + Chromatic
      expect(button).toMatchSnapshot();
    });
  });
});

test("Button with icon", () => {
  const { getByText } = render(
    <Button>
      <Button.Icon icon="search" />
      Click me!
    </Button>,
  );
  const button = getByText("Click me!").element();
  // Snapshots are not the best solution, it woould be much better to set up Storybook + Chromatic
  expect(button).toMatchSnapshot();
});
