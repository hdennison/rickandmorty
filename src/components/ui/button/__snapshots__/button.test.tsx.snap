// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button with icon 1`] = `
<button
  class="flex items-center gap-2 cursor-pointer px-4 py-2 rounded-md justify-center bg-[#ddd] border-[#333] hover:bg-[#ccc]"
>
  <svg
    class="lucide lucide-search"
    fill="none"
    height="16"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="11"
      cy="11"
      r="8"
    />
    <path
      d="m21 21-4.3-4.3"
    />
  </svg>
  Click me!
</button>
`;

exports[`variants > default variant 1`] = `
<button
  class="flex items-center gap-2 cursor-pointer px-4 py-2 rounded-md justify-center bg-[#ddd] border-[#333] hover:bg-[#ccc]"
>
  Click me!
</button>
`;

exports[`variants > primary variant 1`] = `
<button
  class="flex items-center gap-2 cursor-pointer px-4 py-2 rounded-md justify-center bg-[#ddd] border-[#333] hover:bg-[#ccc]"
>
  Click me!
</button>
`;

exports[`variants > secondary variant 1`] = `
<button
  class="flex items-center gap-2 cursor-pointer px-4 py-2 rounded-md justify-center border-1 border-[gray] hover:bg-[#f0f0f0] text-[gray]"
>
  Click me!
</button>
`;
